google-cloud-core/src/main/java/com/google/cloud/CivilTime.java

```diff
@@ -0,0 +1,153 @@
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud;
+
+import com.google.common.base.Preconditions;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Represents a time without timezone, e.g. 07:41:35.220000.
+ *
+ * CivilTime holds values for BigQuery TIME
+ * (https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#time-type).
+ */
+public class CivilTime implements Comparable<CivilTime> {
+
+  private static final Pattern FORMAT_REGEXP = Pattern.compile("(\\d\\d):(\\d\\d):(\\d\\d)(?:\\.(\\d+))?");
+  private final int hours;
+  private final int minutes;
+  private final int seconds;
+  private final int nanos;
+
+  private CivilTime(int hours, int minutes, int seconds, int nanos) {
+    Preconditions.checkArgument(hours >= 0 && hours < 24, "Invalid hours: " + hours);
+    Preconditions.checkArgument(minutes >= 0 && minutes < 60, "Invalid minutes: " + minutes);
+    Preconditions.checkArgument(seconds >= 0 && seconds < 60, "Invalid seconds: " + seconds);
+    Preconditions.checkArgument(nanos >= 0 && nanos <= 999_999_999, "Invalid nanos: " + nanos);
+    this.hours = hours;
+    this.minutes = minutes;
+    this.seconds = seconds;
+    this.nanos = nanos;
+  }
+
+  /**
+   * Constructs a new CivilDate instance.
+   *
+   * @param hours must be between [0,23]
+   * @param minutes must be between [0,59]
+   * @param seconds must be between [0,59]
+   * @param nanos must be between [0,999_999_999]
+   */
+  public static CivilTime ofHoursMinutesSecondsNanos(int hours, int minutes, int seconds, int nanos) {
+    return new CivilTime(hours, minutes, seconds, nanos);
+  }
+
+  /** @param time time in the format hh:mm:ss or hh:mm:ss.SSSSSSSSS */
+  public static CivilTime parseTime(String time) {
+    Matcher matcher = FORMAT_REGEXP.matcher(time);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException("Invalid time: " + time);
+    }
+    int hours = Integer.parseInt(matcher.group(1));
+    int minutes = Integer.parseInt(matcher.group(2));
+    int seconds = Integer.parseInt(matcher.group(3));
+    int nanos = 0;
+    if (matcher.groupCount() > 3) {
+      String fractional = matcher.group(4);
+      Integer fractionalNumber = Integer.parseInt(fractional);
+      if (fractional.length() == 3) {
+        nanos = (int) TimeUnit.MILLISECONDS.toNanos(fractionalNumber);
+      } else if (fractional.length() == 6) {
+        nanos = (int) TimeUnit.MICROSECONDS.toNanos(fractionalNumber);
+      } else {
```

Comment:
Actually we should not throw IAE but parse it correctly irrespective of the number of digits..